File: packages/runtime-python/src/eventgen_runtime/adapters/rabbitmq.py
from __future__ import annotations
import json
import uuid
from eventgen_runtime.types import EventEnvelope, PublishResult


class RabbitMQTransportAdapter:
    def __init__(self, url: str, exchange: str, routing_key: str | None = None) -> None:
        self._url = url
        self._exchange = exchange
        self._routing_key = routing_key
        self._connection = None

    async def _get_channel(self):
        if self._connection is None:
            import aio_pika  # type: ignore[import]
            self._connection = await aio_pika.connect_robust(self._url)
        return await self._connection.channel()

    async def publish(self, envelope: EventEnvelope) -> PublishResult:
        import aio_pika  # type: ignore[import]
        channel = await self._get_channel()
        message_id = str(uuid.uuid4())
        body = json.dumps(envelope.__dict__, default=str).encode("utf-8")
        message = aio_pika.Message(
            body=body,
            content_type="application/json",
            message_id=message_id,
            headers={
                "eventId": envelope.event_id,
                "version": envelope.version,
                "traceId": envelope.metadata.trace_id,
            },
        )
        routing_key = self._routing_key or envelope.event_id
        await channel.default_exchange.publish(message, routing_key=routing_key)
        return PublishResult(message_id=message_id)

    async def close(self) -> None:
        if self._connection:
            await self._connection.close()
            self._connection = None

File: packages/runtime-go/adapters/rabbitmq.go
package adapters

import (
	"context"
	"encoding/json"
	"fmt"

	amqp "github.com/rabbitmq/amqp091-go"
	runtime "github.com/eventgen/runtime-go"
	"github.com/google/uuid"
)

type RabbitMQTransportAdapter struct {
	conn       *amqp.Connection
	ch         *amqp.Channel
	exchange   string
	routingKey string // if empty, uses envelope.EventID
}

func NewRabbitMQTransportAdapter(url, exchange, routingKey string) (*RabbitMQTransportAdapter, error) {
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, fmt.Errorf("rabbitmq adapter: dial: %w", err)
	}
	ch, err := conn.Channel()
	if err != nil {
		conn.Close()
		return nil, fmt.Errorf("rabbitmq adapter: channel: %w", err)
	}
	return &RabbitMQTransportAdapter{conn: conn, ch: ch, exchange: exchange, routingKey: routingKey}, nil
}

func NewRabbitMQTransportAdapterWithChannel(conn *amqp.Connection, ch *amqp.Channel, exchange, routingKey string) *RabbitMQTransportAdapter {
	return &RabbitMQTransportAdapter{conn: conn, ch: ch, exchange: exchange, routingKey: routingKey}
}

func (a *RabbitMQTransportAdapter) Publish(ctx context.Context, envelope runtime.EventEnvelope) (*runtime.PublishResult, error) {
	body, err := json.Marshal(envelope)
	if err != nil {
		return nil, fmt.Errorf("rabbitmq adapter: marshal envelope: %w", err)
	}
	msgID := uuid.NewString()
	rk := a.routingKey
	if rk == "" {
		rk = envelope.EventID
	}
	err = a.ch.PublishWithContext(ctx, a.exchange, rk, false, false, amqp.Publishing{
		ContentType: "application/json",
		MessageId:   msgID,
		Body:        body,
		Headers: amqp.Table{
			"eventId": envelope.EventID,
			"version": envelope.Version,
			"traceId": envelope.Metadata.TraceID,
		},
	})
	if err != nil {
		return nil, fmt.Errorf("rabbitmq adapter: publish %s: %w", envelope.EventID, err)
	}
	return &runtime.PublishResult{MessageID: msgID}, nil
}

func (a *RabbitMQTransportAdapter) Close() error {
	if err := a.ch.Close(); err != nil {
		return err
	}
	return a.conn.Close()
}
